### ISSUE TITLE
compiler: Lightning-indexer top-k selection op (DSA, GLM indexer)

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#97]
## Context
floe `deepseek_v4/forward.py::IndexerScorer`/`LightningIndexer` and glm53flash `Glm53Indexer`: each query scores the compressed entries and keeps the top `index_topk` for attention. Selection is **data-dependent**, but at decode the output count is static (topk) — fits the static-task-grid + runtime-indirection pattern (#94).

## Semantics
- `indexer_scores`: small per-(b,head) scoring task over compressed entries (fixed upper bound = capacity/m, masked by per-row valid length)
- `index_topk`: fixed-count selection producing an indirection table consumed by the attention scores/values tasks via #94
- HCA variant: no indexer (attend all m=128 entries) — same ops, `topk = all`

## Work
selection kernel (bitonic/topk over ≤ ~1k entries per head) as a device template; ordering: indexer phase must complete before the attention phase (phase barrier already provides this)

## Validation
oracle: eager top-k indices match (ties resolved deterministically); logits parity with and without indexer truncation at topk=capacity (should be identical).

Labels: arch:ops, model:deepseek-v4, model:glm-5.3

## Implementation sketch

**Backend**: Triton. Two task kinds: (1) `indexer_scores` — per-row scoring over ≤ S/m entries (≤ ~1k at S=4096/CSA), one block; (2) `index_topk` — in-block fixed-count selection (iterative max-k or bitonic over registers/SMEM), k = `index_topk`, writing an i32 indirection table consumed by #95's scores/values via #94.

**Arithmetic** (fp32 scoring, per floe `IndexerScorer`/`LightningIndexer`):
```
scores = relu(q_i · c_jᵀ) · index_head_dim^−½      # per indexer head, ReLU (not softmax)
w_h    = weights_proj(x) · n_heads^−½               # per-head mixing weights
s_j    = Σ_h scores_{h,j} · w_h                     # fused head mix
top_idx = topk_k(s_j)                               # deterministic tie-break (lowest j)
bias_j = s_j (normalized) → block_bias for the selected entries in #95/#96
```
**Dtypes**: indexer q/entries bf16 (`.cg`), scoring + mixing fp32, indices **i32** `[B, k]`. HCA variant: `topk = all` (attend every m=128 entry) — same tasks, selection skipped.

**Comment by xzyaoi** [comment]
## Status: implemented — branch `issue-97-indexer-topk`, PR open (CPU lane green)

Vertical slice per the issue sketch: two task kinds, capture → lowering → reference exec → Triton device templates → CPU oracle suite.

### What landed (5 commits)
- **IR + capture**: `OP_INDEXER_SCORES` / `OP_INDEX_TOPK` kinds; `ops.indexer_scores(q, entries, mix_w)` records the ReLU scoring + fused head mix (`scale = head_dim**-0.5`, f32 scoring/mixing, bf16-streamable q/entries `.cg`); `ops.index_topk(scores, valid_counts, k)` records the fixed-count selection producing the **i32 indirection table `[B, k]`** + normalized block_bias `[B, k]` consumed by #95/#96 via #94 (RAW hazard → phase barrier ordering, verified in-suite).
- **Semantics pinned** (attributes + numerical contract): `tie_break="lowest_index"`, `score_dtype=f32`, `index_dtype=i32`, `nan_policy=exclude`; candidates at/beyond the per-row valid count are never observed (masked loads — uninitialized tail may hold NaN); slots beyond a row's valid count fill `-1/0.0` (ragged candidate counts, `k > valid_count` supported); `bias = s_j / ||s_valid||_2`; `1 <= k <= capacity` enforced at capture. HCA variant = `k = capacity` (same ops, selection trivial).
- **Lowerings**: `indexer_scores` — one task per (batch, 64-entry tile), per-task reads = row's full `[H, D]` query block + entry tile + mix row; `index_topk` — one task per batch row (M ≤ ~1k sweep), reads score row + i32 valid count. Both on the 256-thread contract, registered in `LOWERINGS` + `TEMPLATE_NAMES`.
- **Reference exec**: fp64 tile-exact bodies; top-k by **rank-by-comparison-counting** (`rank(j) = #{valid i: s_i > s_j} + #{valid i < j: s_i == s_j}`) so the reference is tile-exact against the device contract with no sort.
- **Triton templates**: `_t_indexer_scores`, `_t_index_topk` (single-CTA per row; rank counting in registers — the rank *is* the output slot, deterministic lowest-index tie-break, no shared-memory sort, no scratch).

### Test evidence (`tests/python/test_megakernel_indexer_topk.py` — CPU, 31 passed / 1 skipped, ~0.3 s)
- capture contract + rejection cases (shape/dtype/k mismatches, bf16 storages);
- eager exact oracle (fp64 einsum + `(-score, index)` sort): **indices exact**, bias <1e-9; k=1 and k=M boundaries; tie-heavy coarse-grid inputs; NaN canaries (tail never observed, NaN inside valid prefix excluded from selection and normalizer); vc=0 row; HCA topk=capacity parity (selected set = all candidates, gathered bias = normalized full row — the issue's "logits parity at topk=capacity");
- device-template **mirrors** (numpy fp32 simulation of the exact task decomposition incl. worker strides 1/3/5/7): scores <1e-5 rel, selection indices exact, fill/scatter coverage asserted; full-chain score-multiset agreement <1e-5;
- phase ordering: indexer_scores strictly before index_topk in the schedule.

### Remaining risk
- **CUDA-gated suites unverified in this CPU-only environment** — `_t_indexer_scores`/`_t_index_topk` were validated by CPU mirrors + a CUDA-gated test (`@gpu` skipif), same flagged gap as PRs #88/#113/#114/#115. The rank-scatter store ordering (fill-then-scatter to the same row within one CTA) and bf16 `.cg` paths need one on-device run before merge to production lanes.
- The bias normalizer excludes non-finite valid scores from `||s||` (a NaN inside the prefix would otherwise poison the whole row); if #95/#96 expect normalization over *all* valid entries including NaN-bearing ones, that contract line needs a decision.

**Comment by xzyaoi** [comment]
Superseded by #125 (merged): the Lightning-indexer top-k lane (indexer_scores / index_topk) landed on main. OP_INDEXER_SCORES / OP_INDEX_TOPK present in operator_ir.py on main. Closing per the supersession note on PR #119.
